design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,357 @@
+Map<KsqlServer, Map<String, Map<Integer, LagInfo>>> globalPositionsMap;
+ 
+// Fetch the metadata related to the key
+KeyQueryMetadata queryMetadata = queryMetadataForKey(store, key, serializer);
+ 
+// Acceptable lag for the query
+final long acceptableOffsetLag = 10000;
+
+// Max end offset position
+Map<String, Map<Integer, Long>> maxEndOffsetPerStorePerPartition;
+for (KsqlServer server:  globalPositionsMap) {
+    for (String store: globalPositionsMap.get(server)) {
+        for (Integer partition: globalPositionsMap.get(server).get(store)) {
+            long offset = globalPositionsMap.get(server).get(store).get(partition);
+            maxEndOffsetPerStorePerPartition.computeIfAbsent(store, Map::new);
+            maxEndOffsetPerStorePerPartition.get(store).computeIfAbsent(partition, Map::new);
+            maxEndOffsetPerStorePerPartition.get(store).putIfAbsent(partition, -1);
+            long currentMax = maxEndOffsetPerStore.get(store).get(partition);
+            maxEndOffsetPerStorePerPartition.get(store).put(partition, Math.max(offset, currentMax);       
+        }
+    }
+}
+
+// Ordered list of servers, in order of most caught-up to least
+List<KsqlNode> nodesToQuery;
+KsqlNode active = queryMetadata.getActiveHost();
+if (aliveNodes.get(active)) {
+  nodesToQuery.add(active)  
+}
+
+// add standbys
+nodesToQuery.addAll(queryMetadata.getStandbyHosts().stream()
+    // filter out all the standbys that are down
+    .filter(standbyHost -> aliveNodes.get(standByHost) != null)
+    // get the lag at each standby host for the key's store partition
+    .map(standbyHost -> new Pair(standbyHostInfo,
+        maxEndOffsetPerStorePerPartition.get(storeName).get(queryMetadata.partition()) - 
+        globalPositionsMap.get(standbyHost).get(storeName).get(queryMetadata.partition()).currentOffsetPosition))
+    // Sort by offset lag, i.e smallest lag first
+    .sorted(Comaparator.comparing(Pair::getRight())
+    .filter(standbyHostLagPair -> standbyHostLagPair.getRight() < acceptableOffsetLag)
+    .map(standbyHostLagPair -> standbyHostLagPair.getLeft())
+    .collect(Collectors.toList()));
+ 
+// Query available nodes, starting from active if up
+List<Row> result = null;
+for (KsqlNode server : nodesToQuery) {
+  try {
+    result = query(store, key, server);
+  } catch (Exception e) {
+    System.err.println("Querying server %s failed", server);
+  }
+  if (result != null) {
+    break;
+  }
+}
+
+if (result == null) {
+  throw new Exception("Unable to serve request. All nodes are down or too far behind");
+}
+
+return result;
+```
+
+We will also introduce a per-query configuration parameter `acceptable.offset.lag`, that will provide 
+applications the ability to control how much stale data they are willing to tolerate on a per query
+basis. If a standby lags behind by more than the tolerable limit, pull queries will fail. 
+This parameter can be configured either as part of the `WITH` clause of CTAS queries or be given as
```

Comment:
Why CTAS? I pull query does not create a table -- a CTAS is a persistent query from my understanding? If you refer to the query that populate the table that is queried, I think it does not make much sense to add a `WITH` clause that specified the `acceptable.offset.lag` -- the populating query does not even know anything about pull queries (and should not know anything about it). What would the purpose be to specify it in a CTAS?

Reply:
The configuration is per table as we want to specify how much lag we want to tolerate on this table hence setting it using the CTAS command . Specifying this on a per pull query basis is more fine grained and I am not sure we need this at the moment. Also, the behavior is unpredictable as different queries on the same table return different results or fail based on what value this config has. We can come back to this once implemented and used in the field.

Reply:
What happens is a user does not specify it in the CTAS (`with` is always optional)? If I do a CTAS, I don't know if anybody else will do pull queries against it? Or maybe later on, I change my mind and want to query it? If it's not specified on the pull query either, there must be some server default config.

How much (more) work would it be, to do it per-query from the beginning on. Seems that a per-query model is the most natural and useful. To me, it seems a per-pull-query model would not be more complicated to implement than a per-table model?

Reply:
It's not about more work, it's more the issue of different pull queries on the same table having different behavior. We can do both as well. Tracked in issue #4276

Reply:
Why would that be an issue? If different users specify different "max lag", they would both still get the same answer if the system itself does not change: both query either the active of the most up to date standby -- the only difference could be that one does _not_ get an answer because it's "mag lag" requirement is stricter -- but that should be ok (contrary, it seem to be desired)